test/org/centos/contra/Infra/Utils.groovy

```diff
@@ -0,0 +1,347 @@
+        resource_group_type: "beaker"
+        resource_definitions:
+          - role: "bkr_server"
+            whiteboard: "Dynamically provisioned"
+            recipesets:
+              - name: "hello"
+                distro: "RHEL-6.5"
+                arch: "x86_64"
+                variant: "a"
+                count: 1
+                hostrequires:
+                  - b: "c"
+                    a: "b"
+'''])
+
+    }
+
+    def "openstack topology is created and is correct"() {
+        getPipelineMock("libraryResource")(_) >> {
+            return '''${index}_${providerType}:
+  topology:
+    topology_name: "os-single-new"
+    resource_groups:
+      - resource_group_name: "os-server-new"
+        resource_group_type: "openstack"
+        resource_definitions:
+          - name: "${name}"
+            role: "os_server"
+            flavor: "${instanceType}"
+            image: "${image}"
+            count: 1
+            <% if (keyPair) out.print "keypair: ${keyPair}" %>
+            <% if (fipPool) out.print "fip_pool: ${fipPool}" %>
+            <% if (network) out.print "networks:\\n              - ${network}" %>
+            <% if (autoIP) out.print "auto_ip: ${autoIP}"%>
+        credentials:
+          filename: "openstack.creds"
+          profile: ci-rhos
+'''
+        }
+        def instance = infraUtils.createOpenstackInstances(config.infra.provision.cloud.openstack as HashMap)[0]
+        def instance_name = instance.getNameWithUUID()
+
+        when:
+        infraUtils.generateTopology(instance, 1, "somedir")
+
+        then:
+        noExceptionThrown()
+        1 * getPipelineMock("writeFile")([file: "PinFile", text: """1_openstack:
+  topology:
+    topology_name: "os-single-new"
+    resource_groups:
+      - resource_group_name: "os-server-new"
+        resource_group_type: "openstack"
+        resource_definitions:
+          - name: "$instance_name"
+            role: "os_server"
+            flavor: "m1.small"
+            image: "CentOS-7-x86_64-GenericCloud-1612"
+            count: 1
+            keypair: ci-factory
+            fip_pool: 10.8.240.0
+            networks:
+              - atomic-e2e-jenkins-test
+            
+        credentials:
+          filename: "openstack.creds"
+          profile: ci-rhos
+"""])
+    }
+
+    def "key files can be generated based on provider type"() {
+        when:
+        infraUtils.createKeyFile('aws')
+        infraUtils.createKeyFile('beaker')
+        infraUtils.createKeyFile('openstack')
+        then:
+        1 * getPipelineMock("file.call").call(['credentialsId': "aws.creds", 'variable': 'CREDENTIAL_FILE_NAME'])
+        1 * getPipelineMock("file.call").call(['credentialsId': "beaker.creds", 'variable': 'CREDENTIAL_FILE_NAME'])
+        1 * getPipelineMock("file.call").call(['credentialsId': "openstack.creds", 'variable': 'CREDENTIAL_FILE_NAME'])
+
+    }
+
+    def "ssh keys can be generated for a given provider"() {
+        when:
```

Comment:
Tests should do one thing at a time. I know it can be tedious but this should be split into 3 different tests. Also, looking at the method being tested, i see some logic that is not being accounted for in test. Such as, if `env.SSH_PASSPHRASE` is a thing, or if the `container` closure is getting called

Reply:
@flowerncsu did you see the second part of this comment?

Reply:
Ah, missed that, let me go back and look